### PR TITLE
fix(rslint_parser): delete expressions validation

### PR DESCRIPTION
**Pull request by leops**
## Summary

This adds a number of diagnostics around delete expressions in order to forbid the deletion of private members, single identifiers, and disallow unparenthesized unary expressions in the left hand side of exponentation expressions

### CODE REVIEW

**Review by MichaReiser — CHANGES_REQUESTED** [review]
Nice! And wow, 200 test cases that are now fixed. 

Did you check if it fixes all `delete` expression tests? 

Please add some more tests for some of the edge cases and ensure that any "illegal" expression gets converted to an unknown node. Other than that, looks good. 

PS: I'll replace all the expression specific flags inside of `ParserState` with an `ExpressionContext` that is another bitflag but must be explicitly passed to all expression functions. The main benefit of it is that it becomes clearer when flags get reset, that they don't need to be reset when entering a function or scope (because they're expression specific), and overall simplify reasoning about the code. You don't need to worry about that. I'll deal with merging the changes.

File: crates/rslint_parser/src/syntax/expr.rs [hunk 1]
```diff
@@ -880,27 +922,30 @@ fn parse_paren_or_arrow_expr(p: &mut Parser, can_be_arrow: bool) -> ParsedSyntax
 
             p.error(err);
         } else {
-            if params_marker.is_none() {
-                // Rewind the parser so we can reparse as formal parameters
-                p.rewind(checkpoint);
-                parse_arrow_function_parameters(p, SignatureFlags::empty())
-                    .or_add_diagnostic(p, js_parse_error::expected_parameters);
-            }
+            return p.with_state(InDeleteExpression(false), |p| {
```
[the remark this answers is not in the document]
Comment (leops) [reply crates/rslint_parser/src/syntax/expr.rs:925]: Clearing the flags here isn't actually required as I think any expression within the function body would pass through `parse_expr_or_assignment` and clear it anyway, this was more about keeping the ParserState in a semantically meaningful state (it sounds illogical that the parser could be "in a delete expression" while processing a statement list).
The other calls to `parse_arrow_body` are also only accessible by re-entering `parse_expr_or_assignment` but I did not guard them since this control flow is happening only in `expr.rs` functions calling each other.
For simplicity and keeping everything in line I could remove this guard, it doesn't actually break anything.

File: crates/rslint_parser/src/syntax/expr.rs [hunk 2]
```diff
@@ -144,6 +145,11 @@ pub(super) fn parse_literal_expression(p: &mut Parser) -> ParsedSyntax {
 
 /// Parses an expression that might turn out to be an assignment target if an assignment operator is found
 pub(crate) fn parse_expr_or_assignment(p: &mut Parser) -> ParsedSyntax {
+    p.with_state(InDeleteExpression(false), parse_expr_or_assignment_no_reset)
+}
+
+/// Parse an expression without clearing the `in_delete_expression` flag
+fn parse_expr_or_assignment_no_reset(p: &mut Parser) -> ParsedSyntax {
```
[the remark this answers is not in the document]
Comment (leops) [reply crates/rslint_parser/src/syntax/expr.rs:152]: The expectation is that clearing the delete expression flag is the default behavior in every case except specifically for parenthesized. I could rename `parse_expr_or_assignment_no_reset` to `parse_expr_or_assignment_parenthesized` but then calling it from `parse_expr_or_assignment` would be awkward.
With regards to the spec, the actual behavior described in https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#sec-delete-operator-static-semantics-early-errors would require introspecting the nodes to check for certain patterns, which is not easily done how we currently emit events, so I implemented the checks in a different way that has the same observable effects. To answer the question about test cases, this fixes all of them *but one* which is `delete (ident);` as that would require either actually introspecting the nodes or adding additional state besides tracking whether the parser is in a delete expression (specifically, this flag would need to be decoupled into two "forbid identifier expression" and "forbid private name" flags)

File: crates/rslint_parser/src/syntax/expr.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/rslint_parser/src/syntax/expr.rs]: I see. My refactoring of `ExpressionContext` should then help to make it clear that this is not needed.

File: crates/rslint_parser/src/syntax/expr.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/rslint_parser/src/syntax/expr.rs:152]: Shouldn't it be possible to detect `delete(ident)` by using lookahead inside the parenthesized parser method? 

```
let inner = parse_expr_or_assignment(p);

if p.state.in_delete_expr() && inner.kind() == Some(JS_IDENTIFIER_EXPRESSION) { // add diagnostic 
}
```

Or is the problem that it would then flag `delete this.test[a + (x)]` as well?

[the remark this answers is not in the document]
Comment (leops) [reply crates/rslint_parser/src/syntax/expr.rs:152]: That would work correctly for the indexing case because the index is a new expression (calls `parse_expr_or_assignment`) but it would however flag `delete (a).b;` incorrectly

[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/rslint_parser/src/syntax/expr.rs:152]: How does this differ from `delete (this.#a).b:`?

[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/rslint_parser/src/syntax/expr.rs:152]: I believe the way we can fix this is by unsetting the in delete expr flag for the lhs of the computed binary expression (and all except the last expression inside of the sequence expression). 

This way you know that the delete flag can only be true for the last evaluated expression. I would then probably move the check into the `computed_member_expression` and `parenthesized_expression` to make the assertion a bit clearer (in what circumstances it is relevant)

[the remark this answers is not in the document]
Comment (leops) [reply crates/rslint_parser/src/syntax/expr.rs:152]: Actually that was an oversight on my part, `delete (this.#a).b` should be working but doesn't because the left-hand side of the member expression does not re-enter `parse_expr_or_assignment`, the delete expression should indeed be cleared for these but due to the left-to-right parsing I'm not sure if it's possible to easily know the member expression being parsed is the terminating one in the chain

[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/rslint_parser/src/syntax/expr.rs:152]: Right, that's the hard part about expressions. There's no sufficient lookahead to be able to tell if it turns out to be one or the other except parsing it and then potentially rewinding. 

The only "hack" I can think of right now is to do reparse the events so that we can mark them as unknown if needed but otherwise do a 1:1 forward for token and nodes. OR taking a checkpoint and rewinding if the `lhs` of a binary expression turns out not to be another member expression and then rewind and re-parse without the `in-delete` flag set. 

Or, we wait until we add a second semantic pass to the parser and add it then when we have the possibility to traverse the AST.

[the remark this answers is not in the document]
Comment (leops) [reply crates/rslint_parser/src/syntax/expr.rs]: I pushed a rewrite of this diagnostic based on `rewrite_events` instead of a ParserState flag. It's probably too much for this specific use case since the diagnostic doesn't actually rewrite any event and just needs to inspect them, but I didn't want to introduce a new function to do an immutable scan of the events in the parser if we end up adding a second analysis pass later on.

File: crates/rslint_parser/src/syntax/expr.rs [hunk 3]
```diff
@@ -1618,18 +1715,62 @@ pub(super) fn parse_unary_expr(p: &mut Parser) -> ParsedSyntax {
...
+impl RewriteParseEvents for DeleteExpressionRewriter {
+    fn start_node(&mut self, kind: JsSyntaxKind, p: &mut Parser) {
+        self.stack.push((p.start(), kind));
+        if kind != JS_PARENTHESIZED_EXPRESSION {
+            self.nesting_level += 1;
+        }
+    }
+
+    fn finish_node(&mut self, p: &mut Parser) {
+        if let Some((m, kind)) = self.stack.pop() {
```
Comment (MichaReiser) [inline crates/rslint_parser/src/syntax/expr.rs:1745]: Nit: You can use `unwrap` here. Not having a node in the stack would mean that the stack got corrupted somewhere and that the create diagnostics might be all wrong.

File: crates/rslint_parser/src/syntax/expr.rs [hunk 4]
```diff
@@ -1618,18 +1715,62 @@ pub(super) fn parse_unary_expr(p: &mut Parser) -> ParsedSyntax {
...
+            if kind != JS_PARENTHESIZED_EXPRESSION {
```
Comment (MichaReiser) [inline crates/rslint_parser/src/syntax/expr.rs:1746]: What about sequence expressions: `delete a, b.c`?

File: crates/rslint_parser/src/syntax/expr.rs [hunk 5]
```diff
@@ -1618,18 +1715,62 @@ pub(super) fn parse_unary_expr(p: &mut Parser) -> ParsedSyntax {
...
+                && StrictMode.is_supported(p)
+            {
+                p.error(
+                    p.err_builder("the target for a delete operator cannot be a single identifier")
+                        .primary(node.range(p), ""),
+                );
+                node.change_to_unknown(p);
+            }
+
+            if self.nesting_level == 1 && kind == JS_PRIVATE_NAME {
```
Comment (MichaReiser) [inline crates/rslint_parser/src/syntax/expr.rs:1763]: Is it necessary to verify the parent to ensure it's a `ComputedMemberExpression`. For example, what if `delete this.#a in b`. My understanding is, that this has a nesting level of ` because the private name is nested inside the binary `in` expression.

File: crates/rslint_parser/src/syntax/expr.rs [hunk 6]
```diff
@@ -1606,9 +1647,65 @@ pub(super) fn parse_unary_expr(p: &mut Parser) -> ParsedSyntax {
...
+        // delete obj?.#member;
+        // delete obj?.inner.#member;
+
+        // test_err unary_delete_parenthesized
+        // delete (ident);
+        // delete ((ident));
+        // delete (obj.#member);
+        // delete (func().#member);
+        // delete (obj?.#member);
+        // delete (obj?.inner.#member);
```
Comment (MichaReiser) [inline crates/rslint_parser/src/syntax/expr.rs:1693]: Amazing with how many test cases you came up with!

**Review by MichaReiser — COMMENTED** [review]
Looks good to me. I've two edge cases that might need handling but it should be possible with what we have now. 

Would you mind rebasing onto main so that we can see how this change affects performance?

File: crates/rslint_parser/src/syntax/expr.rs [hunk 7]
```diff
@@ -1618,18 +1715,62 @@ pub(super) fn parse_unary_expr(p: &mut Parser) -> ParsedSyntax {
...
+
+            if self.nesting_level == 0
+                && kind == JS_IDENTIFIER_EXPRESSION
+                && StrictMode.is_supported(p)
+            {
+                p.error(
+                    p.err_builder("the target for a delete operator cannot be a single identifier")
+                        .primary(node.range(p), ""),
+                );
+                node.change_to_unknown(p);
```
Comment (MichaReiser) [inline crates/rslint_parser/src/syntax/expr.rs:1760]: You're rewriting the tree ;)

File: crates/rslint_parser/src/syntax/expr.rs (hunk 4 above)
Comment (leops) [reply crates/rslint_parser/src/syntax/expr.rs]: I pushed a new version of this algorithm that takes advantage of the single identifier / private name being in the trailing position of the expression, so this now works correctly with both sequence expressions and the newly supported `#key in obj` expressions. It would need some logic to handle TypeScript `as` casts such as `delete (obj.key as any)` (which is not not a valid construct, TS emits an error for that)

**Comment by ematipico** [comment]
!bench

**Review by MichaReiser — APPROVED** [review]
Looks good to me. We can solve the typescript issue when working on typescript parsing (if we not decide to move it out entirely into a second check phase).
